### PR TITLE
CASSANDRA-18370 Avoid loading the preferred IP for BulkLoader streaming

### PR DESCRIPTION
**Pull request by maedhroz**
patch by Caleb Rackliffe; reviewed by ? for CASSANDRA-18370

Co-authored-by: Caleb Rackliffe <calebrackliffe@gmail.com>
Co-authored-by: Jon Meredith <jonmeredith@apache.org>

### CODE REVIEW

**Comment by maedhroz** [comment]
https://app.circleci.com/pipelines/github/maedhroz/cassandra?branch=CASSANDRA-18370

File: src/java/org/apache/cassandra/streaming/StreamConnectionFactory.java [hunk 1]
```diff
@@ -26,4 +26,18 @@
...
+     * CASSANDRA-17663 moves calls to SystemKeyspace.getPreferredIP() outside of any threads
+     * that are regularly interrupted.  However the streaming subsystem is also used
+     * by the bulk loader tool, which does not have direct access to the local tables
+     * and uses the client metadata/queries to retrieve it.
+     *
+     * @return true if SystemKeyspace.getPreferredIP() should be used when connecting
+     */
+    default boolean supportsPreferredIp()
+    {
+        return true;
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/streaming/StreamConnectionFactory.java:41]: Note: Setting this to false fails `RepairErrorsTest`, as it should.

File: src/java/org/apache/cassandra/streaming/async/NettyStreamingMessageSender.java [hunk 2]
```diff
@@ -238,7 +238,7 @@ public void sendMessage(StreamMessage message)
                 logger.debug("{} Sending {}", createLogTag(session, null), message);
 
             // Supply a preferred IP up-front to avoid trying to get it in the executor thread, which can be interrupted.
-            OutboundConnectionSettings templateWithConnectTo = template.withConnectTo(template.connectTo());
+            OutboundConnectionSettings templateWithConnectTo = factory.supportsPreferredIp() ? template.withConnectTo(template.connectTo()) : template;
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/streaming/async/NettyStreamingMessageSender.java:241]: Note: This path is tested in `SSTableLoaderEncryptionOptionsTest`.

**Review by jonmeredith — APPROVED** [review]

**Comment by maedhroz** [comment]
Committed as https://github.com/apache/cassandra/commit/853ae8c84049be875921a40c9d5924724cc72792

**Pull request closed without merging**